### PR TITLE
Fixed image resize flashing unexpected size

### PR DESCRIPTION
**Pull request by mlewand**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Fix: Fixed image resize behavior upon short clicking a handle without dragging. Image will no longer became full width, nor will it briefly flash an unexpected size. Closes ckeditor/ckeditor5#5189 and closes ckeditor/ckeditor5#5195.

MINOR BREAKING CHANGE: Resizer options object now also takes an editor instance.

---

### Additional information

* **note** Since [ResizerOptions](https://ckeditor.com/docs/ckeditor5/latest/api/module_widget_widgetresize-ResizerOptions.html) now holds editor reference, it is no longer needed to keep [`downcastWriter`](https://ckeditor.com/docs/ckeditor5/latest/api/module_widget_widgetresize-ResizerOptions.html#member-downcastWriter) so I'd also like to remove this property. I find it to be a minor breaking change, please confirm that.

Short overview of the problems:

* flickering image at the first render (#5189)
	The issue is caused by the fact that we played a little against the UI with the initial implementation.

	The resizer was applying the new `width` styles **directly** to the `img` DOM element, skipping view abstraction. What happened there was that imageresize plugin added a `img.image_resized` class to the `img`'s **view**, which gives it a `100%` width. Since the view knew nothing about the width attribute, it did not add it while re-rendering `img`. Thus was getting full width for a brief moment, until the DOM img was modified once again directly.

	Now implementation works with view, that required me to pass `editor` instance to the resizer so I can request for a view writer.

* committing full image width, when end user only clicks the handle (but does not move the cursor) (#5195)
	The problem is caused by the fact that `commit()` accesses an uninitialized state. And this occurs when user clicks the resize and releases the cursor but does not trigger `mousemove` event, thus the state is never initialized.

	It could be worked around in the imageresize plugin itself (just checking the state here), but the fix should be in the core and prevent commit from being issued.

Subprs:
* ckeditor/ckeditor5-image/pull/334

### CODE REVIEW

File: src/widgetresize/resizer.js [hunk 1]
```diff
@@ -71,6 +79,14 @@ export default class Resizer {
 		this.decorate( 'cancel' );
 		this.decorate( 'commit' );
 		this.decorate( 'updateSize' );
+
+		this.on( 'commit', event => {
+			// State might not be initialized.In this case prevent further handling and make sure that resizer is cleaned up (#5195).
```
Comment (Reinmar) [inline src/widgetresize/resizer.js]:
```suggestion
			// State might not be initialized yet. In this case, prevent further handling and make sure that the resizer is cleaned up (#5195).
```

File: src/widgetresize/resizer.js [hunk 2]
```diff
@@ -62,6 +62,14 @@ export default class Resizer {
 		 */
 		this._domResizerWrapper = null;
 
+		/**
+		 * View to a wrapper of an element controlled by this resizer.
+		 *
+		 * @private
+		 * @type {module:engine/view/element~Element|null}
+		 */
+		this._resizerWrapperView = null;
```
Comment (Reinmar) [inline src/widgetresize/resizer.js]: The "view" doesn't mean that it's a view but it comes from the view so it's an adjective and should be the first word. Just like in the variable which you actually use below:

```js
this._resizerWrapperView = viewResizerWrapper;
```

Next to the `viewResizerWrapper` you have a `domResizerWrapper`.

Comment (Reinmar) [reply src/widgetresize/resizer.js]: Please also fix the description of this property.

**Comment by Reinmar** [comment]
> * I find it to be a minor breaking change, please confirm that.

Yup, that's a minor.

**Comment by Reinmar** [comment]
Unfortunately, I can see a significant performance drop after these changes.

This is `master`:

image

This is with the proposed changes:

image

The first task just got 3x longer. It's clearly visible that it contains a lot more calls.

**Comment by mlewand** [comment]
We just had a discussion with @reinmar about this issue. Here are following actions:

* widget resizer will continue to use views as proposed in this PR
* we'll hide the image toolbar during the image resize
* we'll fix an [issue where ui is forced to be re-rendered if no meaningful mutations were recorded](https://github.com/ckeditor/ckeditor5/issues/5600#issuecomment-564507011)

More details:

The performance hit comes from the fact that now as the code use CKE5 views for DOM manipulation (does not do changes in the DOM directly), there are two [`View#change()`](https://ckeditor.com/docs/ckeditor5/16.0.0/api/module_engine_view_view-View.html#function-change) requests. Each of these change requests finishes with a [`render` event](https://github.com/ckeditor/ckeditor5-engine/blob/28c6f90d62ad33fd1b1199f7a45821ddc546c2d3/src/view/view.js#L479). And on `render` event the image inline toolbar is repositioned - this is where the performance penalty comes from.

On a side note there's also slightly related problem, where changes in [our size UI](https://github.com/ckeditor/ckeditor5-widget/blob/24767b6/src/widgetresize/resizer.js#L482-L493) are causing extra mutations. That's tracked in #5600.

**Comment by mlewand** [comment]
This PR is blocked by [ckeditor/ckeditor5#5964](https://github.com/ckeditor/ckeditor5/issues/5964) as it needs to utilize `WidgetToolbarRepository` forceDisabled() method to hide the toolbar during the resize. Otherwise there are visible performance regression.

**Comment by Reinmar** [comment]
I tested the new disable/enable API as mentioned in [https://github.com/ckeditor/ckeditor5-widget/pull/113](https://github.com/ckeditor/ckeditor5-widget/pull/113). I think you can update this PR now.

**Comment by mlewand** [comment]
Now resizer hides widget toolbar during the resize. I made a soft check on `WidgetToolbarRepository` plugin as it may or may not be available.

Also the actual toolbar hide logic was hooked to Resizer's [`begin`](https://ckeditor.com/docs/ckeditor5/latest/api/module_widget_widgetresize_resizer-Resizer.html#function-begin), `commit` and `cancel` methods instead of mouse listener code directly, so that it works if resizer is operated using the API.

**Comment by FilipTokarski** [comment]
Toolbar works fine now.

There is a strange issue with resizing an image that is in `Fancy editor (padding inside figure)` and has `width: not set`. If you try to resize it veery slowly it suddenly jumps and shrinks just a bit. Sometimes jumps once again when releasing the mouse button. It happens only on the first attempt.
Checked it on Chrome and FF / Win10 + mac.

img_resize__err

**Comment by Mgsy** [comment]
☝️ same in Safari, however, I didn't notice weird behaviour after releasing the mouse button.

**Comment by Reinmar** [comment]
Yeah, the logic for handling an image without a width is tricky and can manifest like that. We had to take some shortcuts there – basically, we estimate the real width. 

Thanks for checking this PR. I'm merging it.

**Pull request merged**